### PR TITLE
fix: add optimize statistic ddl support

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

fix: add optimize statistic ddl support

1. Add new ddl `optimize table table_name  statistic` to calculate table statistics, such as estimate table column distinct values.
2. add `fuse_statistic` table to output some table statistics.
3. unit test cases.

Closes https://github.com/datafuselabs/databend/issues/8465

### CODE REVIEW

**Comment by BohuTANG** [comment]
@lichuang
 Could you add more information to the PR summary? Thanks.
The Databend community expects more from this summary to understand what this PR does.

**Comment by BohuTANG** [comment]
For the new SQL syntax, we need a stateless test for it.

**Comment by BohuTANG** [comment]
> For the new SQL syntax, we need a stateless test for it.

FYI @lichuang

File: tests/logictest/suites/base/09_fuse_engine/09_0008_fuse_optimize_table [hunk 1]
```diff
@@ -45,6 +45,16 @@ select segment_count,block_count from fuse_snapshot('db_09_0008', 't') limit 1;
 ----
 1 1
 
+statement ok
+optimize table `t` statistic;
+
+statement query II
+select segment_count,block_count from fuse_snapshot('db_09_0008', 't') limit 1;
```
Comment (BohuTANG) [inline tests/logictest/suites/base/09_fuse_engine/09_0008_fuse_optimize_table:52]: This sql will check what? It seems not related to statistic

Comment (BohuTANG) [reply tests/logictest/suites/base/09_fuse_engine/09_0008_fuse_optimize_table:52]: This will work:
1. check the statistic
2. optimize table `t` statistic;
3. check the statistic again, make the step2 works

**Comment by lichuang** [comment]
@dantengsky @sundy-li

File: src/query/storages/table-meta/src/meta/v1/snapshot.rs [hunk 2]
```diff
@@ -68,6 +70,7 @@ impl TableSnapshot {
         summary: Statistics,
         segments: Vec<Location>,
         cluster_key_meta: Option<ClusterKey>,
+        table_statistics_location: Option<String>,
```
Comment (sundy-li) [inline src/query/storages/table-meta/src/meta/v1/snapshot.rs:73]: cc @dantengsky   Do we need this? Maybe we could store this using a fixed location by table id?

File: src/query/storages/fuse/fuse/src/fuse_table.rs [hunk 3]
```diff
@@ -149,6 +150,29 @@ impl FuseTable {
         Ok(table_storage_prefix(db_id, table_id))
     }
 
+    pub fn table_snapshot_statistics_format_version(&self, location: &String) -> Result<u64> {
+        Ok(TableMetaLocationGenerator::snapshot_version(location))
```
Comment (dantengsky) [inline src/query/storages/fuse/fuse/src/fuse_table.rs:154]: is `snapshot_statistics_version` supposed to be used here?

File: src/query/storages/fuse/fuse/src/fuse_table.rs [hunk 4]
```diff
@@ -149,6 +150,29 @@ impl FuseTable {
         Ok(table_storage_prefix(db_id, table_id))
     }
 
+    pub fn table_snapshot_statistics_format_version(&self, location: &String) -> Result<u64> {
```
Comment (dantengsky) [inline src/query/storages/fuse/fuse/src/fuse_table.rs:153]: seem that returning a plain `u64` is also ok

File: src/query/storages/table-meta/src/meta/v1/snapshot.rs (hunk 2 above)
Comment (dantengsky) [reply src/query/storages/table-meta/src/meta/v1/snapshot.rs:73]: For the evolution of the table statistics file format, I suggest keeping the location specified explicitly (if I get this pr correct, the version of the statistics file is detected by parsing its location)

File: src/query/storages/fuse/fuse/src/fuse_table.rs (hunk 3 above)
Comment (lichuang) [reply src/query/storages/fuse/fuse/src/fuse_table.rs]: yes, in `read_table_snapshot_statistics`

File: src/query/service/tests/it/storages/fuse/operations/optimize.rs [hunk 5]
```diff
@@ -19,25 +19,38 @@ use crate::storages::fuse::table_test_fixture::append_sample_data;
...
 
 #[tokio::test]
 async fn test_fuse_snapshot_optimize_purge() -> Result<()> {
-    do_purge_test("explicit pure", "purge").await
+    do_purge_test("explicit pure", "purge", 1, 0, 1, 1, 1).await
+}
+
+#[tokio::test]
+async fn test_fuse_snapshot_optimize_statistic() -> Result<()> {
+    do_purge_test("explicit pure", "statistic", 3, 1, 2, 2, 2).await
```
Comment (dantengsky) [inline src/query/service/tests/it/storages/fuse/operations/optimize.rs:37]: are **3** snapshots supposed to be left there, after one insertion and a `optimize stats` operation?

failed to figure out why the number of segments and blocks also increased, would you please add some docs here?

File: src/query/service/src/interpreters/interpreter_table_optimize.rs [hunk 6]
```diff
@@ -112,6 +113,10 @@ impl Interpreter for OptimizeTableInterpreter {
             }
         }
 
+        if do_statistic {
+            table.statistic(self.ctx.clone()).await?;
```
Comment (dantengsky) [inline src/query/service/src/interpreters/interpreter_table_optimize.rs:117]: shall we return here after the statistic has been called?

File: src/query/service/tests/it/storages/fuse/operations/optimize.rs (hunk 5 above)
Comment (lichuang) [reply src/query/service/tests/it/storages/fuse/operations/optimize.rs]: Since `statistic`  operation:
1. generate a new snapshot
2. and does not purge the last snapshot

so,  after `statistic`, snapshot count will be add from 2 to 3, i add a test that after `statistic` operation do a purge, and snapshot will become to 1.

File: src/query/service/src/interpreters/interpreter_table_optimize.rs (hunk 6 above)
Comment (lichuang) [reply src/query/service/src/interpreters/interpreter_table_optimize.rs:117]: OK

**Review by BohuTANG — APPROVED** [review]

**Comment by BohuTANG** [comment]
cc @soyeric128 for documentation. Thanks.
